util/reggen/gen_tock.py

```diff
@@ -0,0 +1,384 @@
+    highest_address = set()  # type: Set[int]
+
+    gen_const_module_params(paramout, block, block.name, block.regwidth,
+                            existing_defines)
+
+    gen_const_interrupts(fieldout, block, block.name, block.regwidth,
+                         existing_defines)
+
+    for rb in block.reg_blocks.values():
+        for x in rb.entries:
+            if isinstance(x, Register):
+                gen_const_register(regout, fieldout, x, block, rnames,
+                                   existing_defines, access_type,
+                                   highest_address)
+                continue
+
+            if isinstance(x, MultiRegister):
+                gen_const_multireg(regout, fieldout, x, block, rnames,
+                                   existing_defines, access_type,
+                                   highest_address)
+                continue
+
+            if isinstance(x, Window):
+                gen_const_window(regout, x, block, rnames, existing_defines,
+                                 access_type, highest_address)
+                continue
+
+    paramstr = paramout.getvalue()
+    paramout.close()
+
+    end = sorted(highest_address)[-1]
+    # End the register file.
+    genout(regout, '(0x{:04x} => @END),\n', end)
+    # Close the "pub <block>" declaration, then the "register_structs!" macro.
+    genout(regout, '}}\n}}\n\n')
+    regstr = regout.getvalue()
+    regout.close()
+
+    # Close the "register_bitfields!" macro.
+    genout(fieldout, '];\n\n')
+    fieldstr = fieldout.getvalue()
+    fieldout.close()
+
+    genout(outfile, '// Generated register constants for {}\n\n', block.name)
+    genout(outfile, '// Built for {}\n',
+           version_stamp.get('BUILD_GIT_VERSION', '<unknown>'))
+    genout(outfile, '// https://github.com/lowRISC/opentitan/tree/{}\n',
+           version_stamp.get('BUILD_SCM_REVISION', '<unknown>'))
+    genout(outfile, '// Tree status: {}\n',
+           version_stamp.get('BUILD_SCM_STATUS', '<unknown>'))
+    tm = int(version_stamp.get('BUILD_TIMESTAMP', 0))
+    dt = datetime.utcfromtimestamp(tm) if tm else datetime.utcnow()
+    genout(outfile, '// Build date: {}\n\n', dt.isoformat())
+
+    if src_copy:
+        genout(outfile, '// Copyright information found in source file:\n')
+        genout(outfile, '// {}\n\n', src_copy)
+    if src_lic is not None:
+        genout(outfile, '// Licensing information found in source file:\n')
+        for line in src_lic.splitlines():
+            genout(outfile, '// {}\n', line)
+        genout(outfile, '\n')
+    genout(outfile, "use kernel::utilities::registers::{{\n")
+    genout(outfile, "    register_bitfields, register_structs, {}\n",
+           ', '.join(sorted(access_type)))
+    genout(outfile, "}};\n")
+
+    outfile.write(indent(paramstr))
+    outfile.write(indent(regstr))
+    outfile.write(indent(fieldstr))
+
+    genout(outfile, '// End generated register constants for {}', block.name)
```

Comment:
nit: This should write an EOL character. Currently, the generated files are missing EOL characters on their last lines.

Reply:
Done.